llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -90,24 +90,35 @@ const Constant *getAliasee(const GlobalAlias *GA) {
 
 template <typename GlobalType>
 void ModuleHashGenerator::calculateGOInfo(const GlobalType *G) {
-  GOInfo Result = calculateDigestAndDependenciesAndContributedToGVs(G);
-  // ContributedToGVs of an object is used to update other objects'
-  // contributions. For example, if the ContributedToGVs of function `foo`  are
-  // global variables `g` and `q`, the function `foo` is in the contributions of
-  // `g` and `q`.
-  // ContributedToGVs[`foo`] = [`g`, `q`]
-  // ====> Contributions[`g`] = [`foo`],
-  //       Contributions[`q`] = [`foo`]
+  GODetailedInfo Result = calculateDigestAndDependenciesAndContributions(G);
+#ifndef NDEBUG
+  LLVM_DEBUG(dbgs() << "\nGO Name:" << G->getName() << "\n");
+  LLVM_DEBUG(Result.dump());
+#endif
+  // Update G's dependencies.
+  GOInfo &GInfo = GOIMap[G];
+  GInfo.InitialDigest = std::move(Result.HashInfo.InitialDigest);
+  GInfo.Dependencies.insert(GInfo.Dependencies.end(),
+                            Result.HashInfo.Dependencies.begin(),
+                            Result.HashInfo.Dependencies.end());
+
+  // Contributions of an object is used to update other objects' Dependencies.
+  // For example, if the Contributions of function `foo`  are global variables
+  // `g` and `q`, the function `foo` is added into the Dependencies of `g` and
+  // `q`.
+  // For example, Contributions[`foo`] = [`g`, `q`]
+  // ====> Dependencies[`g`] = [`foo`],
+  //       Dependencies[`q`] = [`foo`]
+  //       Dependencies[`foo`] = [`g`, `q`]
+  //
+  // The below loop converts Contributions to Dependencies and then just records
+  // Dependencies in the resulting graph.
   for (auto &GO : Result.Contributions) {
     assert(isa<GlobalVariable>(GO) &&
-           "Only global variables can have contributions!");
-    assert(isa<llvm::Function>(G) && "All contributions are functions!");
-    GOIMap[GO].Contributions.emplace_back(G);
+           "Contributions must be global variables!");
+    GOIMap[GO].Dependencies.emplace_back(G);
+    GOIMap[G].Dependencies.emplace_back(GO);
```

Comment:
Isn’t `GOIMap[G]` the same as `GInfo` (assigned on line 99 above)? If `GOIMap` is a map-like structure (as its name suggests) then we’ll be searching the tree for the same object every time round this loop.

Reply:
Fixed in https://github.com/SNSystems/llvm-project-prepo/pull/127/commits/4a8756041970f24b02df38128f15c683126a7ebe.